File: playground_cpp/src/camera.cpp
#include "camera.h"
#include <glm/gtc/matrix_transform.hpp>
#include <cmath>

namespace Camera {

glm::mat4 perspective(float fovY, float aspect, float nearPlane, float farPlane) {
    // Vulkan clip space: Y-down (flip Y), depth [0, 1]
    // This matches the Python playground's perspective() function
    float f = 1.0f / std::tan(fovY / 2.0f);

    glm::mat4 m(0.0f);
    m[0][0] = f / aspect;
    m[1][1] = -f;  // Vulkan Y-flip
    m[2][2] = farPlane / (nearPlane - farPlane);
    m[2][3] = -1.0f;
    m[3][2] = (nearPlane * farPlane) / (nearPlane - farPlane);

    return m;
}

glm::mat4 lookAt(glm::vec3 eye, glm::vec3 target, glm::vec3 up) {
    // Standard look-at matrix matching the Python playground
    glm::vec3 f = glm::normalize(target - eye);
    glm::vec3 s = glm::normalize(glm::cross(f, up));
    glm::vec3 u = glm::cross(s, f);

    glm::mat4 m(1.0f);
    m[0][0] = s.x;  m[1][0] = s.y;  m[2][0] = s.z;
    m[0][1] = u.x;  m[1][1] = u.y;  m[2][1] = u.z;
    m[0][2] = -f.x; m[1][2] = -f.y; m[2][2] = -f.z;
    m[3][0] = -glm::dot(s, eye);
    m[3][1] = -glm::dot(u, eye);
    m[3][2] = glm::dot(f, eye);

    return m;
}

CameraMatrices getDefaultMatrices(float aspect) {
    CameraMatrices cam;
    cam.model = glm::mat4(1.0f); // identity
    cam.view = lookAt(DEFAULT_EYE, DEFAULT_TARGET, DEFAULT_UP);
    cam.projection = perspective(
        glm::radians(DEFAULT_FOV), aspect, DEFAULT_NEAR, DEFAULT_FAR);
    return cam;
}

RTCameraData getRTCameraData(float aspect) {
    RTCameraData data;
    glm::mat4 view = lookAt(DEFAULT_EYE, DEFAULT_TARGET, DEFAULT_UP);
    glm::mat4 proj = perspective(
        glm::radians(DEFAULT_FOV), aspect, DEFAULT_NEAR, DEFAULT_FAR);
    data.inverseView = glm::inverse(view);
    data.inverseProjection = glm::inverse(proj);
    return data;
}

} // namespace Camera

File: playground_rust/src/camera.rs
//! Camera math: perspective projection and look-at view matrix.
//!
//! All matrices use Vulkan clip-space conventions:
//! - Y is flipped (negative in projection)
//! - Depth range [0, 1]
//! - Column-major storage (glam default)

use glam::{Mat4, Vec3};

/// Create a perspective projection matrix for Vulkan clip space.
///
/// The Y axis is flipped (m[1][1] = -f) to match Vulkan's top-down convention.
/// Depth maps to [0, 1] (not [-1, 1] like OpenGL).
pub fn perspective(fov_y: f32, aspect: f32, near: f32, far: f32) -> Mat4 {
    let f = 1.0 / (fov_y / 2.0).tan();

    // Build column-major matrix matching the Python playground exactly.
    // Column 0: (f/aspect, 0, 0, 0)
    // Column 1: (0, -f, 0, 0)        <- Y-flip for Vulkan
    // Column 2: (0, 0, far/(near-far), -1)
    // Column 3: (0, 0, near*far/(near-far), 0)
    Mat4::from_cols(
        glam::Vec4::new(f / aspect, 0.0, 0.0, 0.0),
        glam::Vec4::new(0.0, -f, 0.0, 0.0),
        glam::Vec4::new(0.0, 0.0, far / (near - far), -1.0),
        glam::Vec4::new(0.0, 0.0, (near * far) / (near - far), 0.0),
    )
}

/// Create a look-at view matrix (column-major).
///
/// Matches the Python playground's `look_at()` function.
pub fn look_at(eye: Vec3, target: Vec3, up: Vec3) -> Mat4 {
    let f = (target - eye).normalize();
    let s = f.cross(up).normalize();
    let u = s.cross(f);

    // Column-major storage matching Python numpy layout:
    // Column 0: (s.x, u.x, -f.x, 0)
    // Column 1: (s.y, u.y, -f.y, 0)
    // Column 2: (s.z, u.z, -f.z, 0)
    // Column 3: (-dot(s,eye), -dot(u,eye), dot(f,eye), 1)
    Mat4::from_cols(
        glam::Vec4::new(s.x, u.x, -f.x, 0.0),
        glam::Vec4::new(s.y, u.y, -f.y, 0.0),
        glam::Vec4::new(s.z, u.z, -f.z, 0.0),
        glam::Vec4::new(-s.dot(eye), -u.dot(eye), f.dot(eye), 1.0),
    )
}

/// Default camera parameters matching the Python playground.
pub struct DefaultCamera;

impl DefaultCamera {
    pub const EYE: Vec3 = Vec3::new(0.0, 0.0, 3.0);
    pub const TARGET: Vec3 = Vec3::ZERO;
    pub const UP: Vec3 = Vec3::new(0.0, 1.0, 0.0);
    pub const FOV_Y_DEG: f32 = 45.0;
    pub const NEAR: f32 = 0.1;
    pub const FAR: f32 = 100.0;

    pub fn view() -> Mat4 {
        look_at(Self::EYE, Self::TARGET, Self::UP)
    }

    pub fn projection(aspect: f32) -> Mat4 {
        perspective(Self::FOV_Y_DEG.to_radians(), aspect, Self::NEAR, Self::FAR)
    }

    pub fn model() -> Mat4 {
        Mat4::IDENTITY
    }
}
